src/argparcel/magic.py

```diff
@@ -0,0 +1,167 @@
+        ),
+        None,
+    )
+    assert class_node is not None
+
+    result: dict[str, str] = {}
+
+    # Walk through the class body to collect field docstrings. We're looking for pairs
+    # of field definition followed by a docstring.
+    for node, next_node in itertools.pairwise(class_node.body):
+        if isinstance(node, ast.AnnAssign) and isinstance(node.target, ast.Name):
+            field_name = node.target.id
+
+            if (
+                isinstance(next_node, ast.Expr)
+                and isinstance(next_node.value, ast.Constant)
+                and isinstance(next_node.value.value, str)
+            ):
+                result[field_name] = next_node.value.value
+
+    return result
+
+
+def _is_type_checking_block(node: ast.AST) -> typing.TypeGuard[ast.If]:
+    """Returns True if the given AST node is an `if TYPE_CHECKING:` block."""
+    if not isinstance(node, ast.If):
+        return False
+
+    test = node.test
+    if isinstance(test, ast.Name):
+        return test.id == "TYPE_CHECKING"
+    if isinstance(test, ast.Attribute):
+        return (
+            isinstance(test.value, ast.Name)
+            and test.value.id == "typing"
+            and test.attr == "TYPE_CHECKING"
+        )
+    return False
+
+
+def _module_globals_including_type_checking(
+    module: types.ModuleType, *, allowed_extra_names: frozenset[str]
+) -> dict[str, typing.Any]:
+    """Get the global namespace for `module`... including imports in TYPE_CHECKING.
+
+    Of any imports identified in TYPE_CHECKING blocks, only names that appear in
+    `allowed_extra_names` will actually be imported.
+    """
+    source = inspect.getsource(module)
+    globalns = dict(vars(module))
+
+    tree = ast.parse(source)
+
+    for node in tree.body:
+        if not _is_type_checking_block(node):
+            continue
+
+        for statement in node.body:
+            if isinstance(statement, ast.Import):
+                for alias in statement.names:
+                    if alias.name not in allowed_extra_names:
+                        continue
+                    mod = importlib.import_module(alias.name)
+                    asname = alias.asname or alias.name
+                    globalns[asname] = mod
+
+            elif isinstance(statement, ast.ImportFrom):
+                modname = statement.module
+                assert modname is not None
+                mod = importlib.import_module(modname)
+                for alias in statement.names:
+                    if alias.name not in allowed_extra_names:
+                        continue
+                    name = alias.name
+                    asname = alias.asname or name
+                    globalns[asname] = getattr(mod, name)
+
+    return globalns
+
+
+def _names_used_in_annotation(cls: type[_typeshed.DataclassInstance]) -> frozenset[str]:
+    """Extract names used in field annotations of `cls`."""
```

Comment:
The point of this function is to avoid running _all_ of the imports in a TYPE_CHECKING block, but filter out just those that are plausibly required.